## question 1A
generate six sentences

In [59]:
A = "I love Management Mathematics"
B = "I hate Management Mathematics"
C = "I love Operations Research"
D = "I hate Operations Research"
E = "I love Manufacturing Data Science"
F = "I hate Manufacturing Data Science"

## Question 1B
Get the list of words in each sentence by calling the split() in strings, and convert the sentence into lowercase.

In [60]:
import numpy as np
wordsA = A.lower().split()
wordsB = B.lower().split()
wordsC = C.lower().split()
wordsD = D.lower().split()
wordsE = E.lower().split()
wordsF = F.lower().split()
print(wordsA)
print(wordsB)
print(wordsC)
print(wordsD)
print(wordsE)
print(wordsF)

['i', 'love', 'management', 'mathematics']
['i', 'hate', 'management', 'mathematics']
['i', 'love', 'operations', 'research']
['i', 'hate', 'operations', 'research']
['i', 'love', 'manufacturing', 'data', 'science']
['i', 'hate', 'manufacturing', 'data', 'science']


## Question 1C
Use the set class to identify the dimensionality of the feature space and print all vocabs.

In [61]:
vocab = set(wordsA)
vocab = vocab.union(set(wordsB))
vocab = vocab.union(set(wordsC))
vocab = vocab.union(set(wordsD))
vocab = vocab.union(set(wordsE))
vocab = vocab.union(set(wordsF))
print(vocab)

{'manufacturing', 'data', 'research', 'operations', 'mathematics', 'love', 'i', 'science', 'management', 'hate'}


## 1D 
To build the one-to-one mapping between features and IDs in the feature space. We use the list() to convert the vocabs into unique features.

In [62]:
vocab = list(vocab)
print(vocab)

['manufacturing', 'data', 'research', 'operations', 'mathematics', 'love', 'i', 'science', 'management', 'hate']


## 1E
declare six 1D arrays for the six sentences and build each word vector by increasing the corresponding feature values

In [63]:
vA = np.zeros(len(vocab), dtype=float)
vB = np.zeros(len(vocab), dtype=float)
vC = np.zeros(len(vocab), dtype=float)
vD = np.zeros(len(vocab), dtype=float)
vE = np.zeros(len(vocab), dtype=float)
vF = np.zeros(len(vocab), dtype=float)

for w in wordsA:
    i = vocab.index(w)
    vA[i] += 1
print(vA)

for w in wordsB:
    i = vocab.index(w)
    vB[i] += 1
print(vB)

for w in wordsC:
    i = vocab.index(w)
    vC[i] += 1
print(vC)

for w in wordsD:
    i = vocab.index(w)
    vD[i] += 1
print(vD)

for w in wordsE:
    i = vocab.index(w)
    vE[i] += 1
print(vE)

for w in wordsF:
    i = vocab.index(w)
    vF[i] += 1
print(vF)

[0. 0. 0. 0. 1. 1. 1. 0. 1. 0.]
[0. 0. 0. 0. 1. 0. 1. 0. 1. 1.]
[0. 0. 1. 1. 0. 1. 1. 0. 0. 0.]
[0. 0. 1. 1. 0. 0. 1. 0. 0. 1.]
[1. 1. 0. 0. 0. 1. 1. 1. 0. 0.]
[1. 1. 0. 0. 0. 0. 1. 1. 0. 1.]


## 1F
Compute the cosine similarity between two vectors and show a 6x6 matrix including these similarity values

In [64]:
import pandas as pd
arr = np.zeros((6,6))
xc = 0
xr = 0
for i in [vA, vB, vC, vD, vE, vF]:
    xr = 0
    for j in [vA, vB, vC, vD, vE, vF]:
        cos = np.dot(i, j) / (np.sqrt(np.dot(i,i)) * np.sqrt(np.dot(j,j)))
        arr[xc][xr] = cos
        xr += 1
    xc += 1
data = pd.DataFrame(arr, columns = ["vA", "vB", "vC", "vD", 'vE', 'vF'])

In [65]:
data

vA        vB        vC        vD        vE        vF
0  1.000000  0.750000  0.500000  0.250000  0.447214  0.223607
1  0.750000  1.000000  0.250000  0.500000  0.223607  0.447214
2  0.500000  0.250000  1.000000  0.750000  0.447214  0.223607
3  0.250000  0.500000  0.750000  1.000000  0.223607  0.447214
4  0.447214  0.223607  0.447214  0.223607  1.000000  0.800000
5  0.223607  0.447214  0.223607  0.447214  0.800000  1.000000

## 1G 
based on the similarity what;s the result you draw?

Based on the cosine similarity matrix provided, we can make several observations:

The cosine similarity between a vector and itself is always 1. This is reflected in the diagonal elements of the matrix (1.000000 in all cases).

The cosine similarity between vectors vA and vB (0.750000) is the same as the cosine similarity between vectors vB and vA. This is because cosine similarity is symmetric.

The cosine similarity between vectors vA and vC (0.500000) indicates that they have a moderate similarity, but not as high as the similarity between vA and vB.

The cosine similarity between vectors vA and vD (0.250000) indicates a lower similarity compared to the previous cases.

The cosine similarity between vectors vA and vE (0.447214) is the same as the similarity between vA and vC. This suggests that vE and vC have a similar degree of similarity to vA.

The cosine similarity between vectors vA and vF (0.223607) is the lowest among all the pairwise similarities involving vA. This indicates that vF is the least similar to vA compared to the other vectors.

The cosine similarity between vectors vD and vE (0.223607) is the same as the similarity between vA and vF. This suggests that vD and vE have a similar degree of similarity to vA, which is lower compared to the similarities involving vB, vC, and vE.



## 2A
please follow the course lecture notes to finish the IR work. Build up the matrix for preparing IR and convert all column vectors into unit vectors as matrix Q.

In [66]:
data = [[0,0,5,6,0,0,0,5,0,0],[6,0,4,5,0,0,0,3,0,4],[3,0,4,3,0,0,5,3,0,4],[0,0,5,3,0,0,2,2,5,3],[1,0,4,4,3,4,3,4,1,4],[0,5,0,4,0,6,3,2,3,1],[1,3,3,3,4,0,0,1,1,0],[1,2,3,2,3,2,1,1,0,3]]
A = pd.DataFrame(data).T
A.columns=["M1", "M2", "M3", "M4", "M5", "M6", "M7", "M8"]
A.index=["determinants", "eigenvalues", "linear", "matrices", "numerical", "orthogonality", "spaces", "systems", "transformations", "vector"]
A

M1  M2  M3  M4  M5  M6  M7  M8
determinants      0   6   3   0   1   0   1   1
eigenvalues       0   0   0   0   0   5   3   2
linear            5   4   4   5   4   0   3   3
matrices          6   5   3   3   4   4   3   2
numerical         0   0   0   0   3   0   4   3
orthogonality     0   0   0   0   4   6   0   2
spaces            0   0   5   2   3   3   0   1
systems           5   3   3   2   4   2   1   1
transformations   0   0   0   5   1   3   1   0
vector            0   4   4   3   4   1   0   3

In [67]:
import math
arr = np.zeros((8, 10))
countx=0
for i in [A["M1"], A["M2"],A["M3"], A["M4"], A["M5"], A["M6"], A["M7"], A["M8"]]:
    county=0
    d = 0
    for j in i:
        d += j**2
    for j in i:
        arr[countx][county] = j/math.sqrt(d)
        county += 1
    countx += 1
Q = pd.DataFrame(arr).T
Q.columns=["M1", "M2", "M3", "M4", "M5", "M6", "M7", "M8"]
Q.index=["determinants", "eigenvalues", "linear", "matrices", "numerical", "orthogonality", "spaces", "systems", "transformations", "vector"]
Q

M1        M2        M3        M4   M5   M6        M7  \
determinants     0.000000  0.594089  0.327327  0.000000  0.1  0.0  0.147442   
eigenvalues      0.000000  0.000000  0.000000  0.000000  0.0  0.5  0.442326   
linear           0.539164  0.396059  0.436436  0.573539  0.4  0.0  0.442326   
matrices         0.646997  0.495074  0.327327  0.344124  0.4  0.4  0.442326   
numerical        0.000000  0.000000  0.000000  0.000000  0.3  0.0  0.589768   
orthogonality    0.000000  0.000000  0.000000  0.000000  0.4  0.6  0.000000   
spaces           0.000000  0.000000  0.545545  0.229416  0.3  0.3  0.000000   
systems          0.539164  0.297044  0.327327  0.229416  0.4  0.2  0.147442   
transformations  0.000000  0.000000  0.000000  0.573539  0.1  0.3  0.147442   
vector           0.000000  0.396059  0.436436  0.344124  0.4  0.1  0.000000   

                       M8  
determinants     0.154303  
eigenvalues      0.308607  
linear           0.462910  
matrices         0.308607  
numerical        0.462910  
orthogonality    0.308607  
spaces           0.154303  
systems          0.154303  
transformations  0.000000  
vector           0.462910

First, we use pandas to create a dataframe, and add the indexes and columns on the dataframe. then, in order to calculate the unit vectors as matrix Q , we calculate the ||X|| for each columns, and calculate the values of each positions.

## 2B
form a unit search vector x whose entries are all 0 except for the three rows corresponding to the search rows.

In [68]:
x = pd.Series([0,0,0,0,0,1/math.sqrt(3),1/math.sqrt(3),0,0,1/math.sqrt(3) ])
x

0    0.00000
1    0.00000
2    0.00000
3    0.00000
4    0.00000
5    0.57735
6    0.57735
7    0.00000
8    0.00000
9    0.57735
dtype: float64

## 2C
Calculate the cosine similarity y=Q.t * X and rank the modules.

In [69]:
y = np.dot(Q.T, x)
y

array([0.        , 0.22866478, 0.56694671, 0.33113309, 0.6350853 ,
       0.57735027, 0.        , 0.53452248])

In [70]:
rank = np.argsort(y)[::-1]
rank += 1
print(rank)

[5 6 3 8 4 2 7 1]
